### ISSUE TITLE
HttpClient, AllowAutoRedirect 

### ISSUE THREAD

**Issue by logwlogw** [dotnet/corefx#33846]
Hello, 

Bug with  AllowAutoRedirect.

I try download many urls (unique) asynchronously (100K -1M)
(200-500 async Tasks) 

I use:
1. static httpclient
2. SocketsHttpHandler
```c#
     var handler = new SocketsHttpHandler
            {
                UseProxy = false,
                ConnectTimeout = TimeSpan.FromMilliseconds(100 * 1000),
                PooledConnectionIdleTimeout = TimeSpan.Zero,
                PooledConnectionLifetime = TimeSpan.Zero,
                ResponseDrainTimeout = TimeSpan.Zero,
                Expect100ContinueTimeout = TimeSpan.Zero,
                UseCookies = false,
                AllowAutoRedirect = true,
                MaxAutomaticRedirections = 1
            };
```

**I 99% CPU  Load on VPS Hosting (One Core). CentOS** and low progress

If I Use  AllowAutoRedirect = false,

**0-10% CPU Load.**

**Comment by rmkerr** [comment]
Hey @logwlogw, thanks for the report! Are a significant proportion of the requests receiving redirects as a response? If so it makes a lot of sense that you would see less throughput, since a typical 4XX redirect response is going to be a whole lot smaller (and thus download a lot quicker) than a normal webpage.

**Comment by logwlogw** [comment]
Thank you very much for your response.

Yes, a lot of 302,303 codes.
~ 20-40%

I wrote my handler 302-303. +  AllowAutoRedirect = false

    public async Task<HttpLibResponse> DownloadUrl(string url)
        {
            int countretry = 0;
            int MaxAutomaticRedirections = 10;


            ReMovedLabel:
            var response = await DownloadUrl_Int2(url).ConfigureAwait(false);
            if (_token.IsCancellationRequested)
            {
                return response;
            }



            if ((response.StatusCode == HttpStatusCode.Moved) || (response.StatusCode == HttpStatusCode.Redirect))
            {
                if (countretry >= MaxAutomaticRedirections)
                {
                    return response;
                }


                countretry++;
                url = response.Location;
                goto ReMovedLabel;
            }


            return response;
        }


      public async Task<HttpLibResponse> DownloadUrl_Int2(string url)
        {
            var response = new HttpLibResponse();
            try
            {
                var uri = UrlCorrect.GetUrlFromString(url);
                if (uri == null) throw new Exception($"Incorrect uri {url}");


                var request = new HttpRequestMessage(HttpMethod.Get, uri);

                using (HttpResponseMessage responseclient = await _client.SendAsync(request, _token).ConfigureAwait(false))
                {
                    using (var responseContent = responseclient.Content)
                    {
                        response.StatusCode = responseclient.StatusCode;
                        response.StatusCodeReason = responseclient.ReasonPhrase;
                        response.Headers = responseclient.Headers.ToString() + responseContent.Headers;

                        response.Content = await responseContent.ReadAsStringAsync().ConfigureAwait(false);

                        if ((response.StatusCode == HttpStatusCode.Moved) || (response.StatusCode == HttpStatusCode.Redirect))
                        {
                            response.Location = responseclient.Headers.Location.ToString();
                        }

                        response.Success = true;
                    }
                }
            }
            catch (TaskCanceledException e)
            {
                if (!_token.IsCancellationRequested)
                {
                    response.TimedOut = true;
                }
                response.Success = false;
                response.StatusCodeReason = e.Message;
            }
            catch (Exception e)
            {
                response.Success = false;
                response.StatusCodeReason = e.Message;
            }
            return response;
        }


if MaxAutomaticRedirections = 0 -All Ok
if MaxAutomaticRedirections = 10 -99% CPU Load


I do not understand the reason: (

On win10, everything works.

**Comment by logwlogw** [comment]
Yes, if I set up in my code MaxAutomaticRedirections = 10.

**Comment by rmkerr** [comment]
The first three settings you are configuring there could certainly be affecting performance, unless all of your requests are to different servers. I would remove those lines completely (which leaves them at the default values) and see if performance is any better. The last option is probably unrelated.

To give some more background on those first three options:

One of the optimizations that we use in `SocketsHttpHandler` is connection pooling. Connections are fairly slow to open, so rather than opening one for each request, we keep a bunch of connections open (aka the connection pool). When a new request is sent, it uses a connection from the pool rather than opening a new one.

Some of our users need to be able to control the lifetime of the connections in the pool, which is why we provide the first three APIs you listed. `PooledConnectionLifetime ` allows users to set the total lifetime for a connection on the pool. Setting it to 0 means connections will immediately become invalid after the first request, totally negating any perf benefits of connection pooling. `PooledConnectionIdleTimeout` is similar, but is based on how long the connection has gone without sending data, rather than the total connection lifetime.

`ResponseDrainTimeout` helps us reuse connections that may still have unread data that was ignored by whoever processed the received response. To avoid creating a new connection, we will often try to read (and discard) all of the data waiting on the connection. This property basically just controls how long we'll try that for, before giving up and creating a new connection.

**Comment by logwlogw** [comment]
Thank you very much for your response.
Simple PerFomance Test.

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;

namespace PerfTest
{
    class Program
    {
        static readonly int MaxTasks = 100;

        static readonly Stopwatch St = new Stopwatch();

        private static readonly string[] Urls = File.ReadAllLines(@"10000urls.txt");
        private static int _urlindex;

        static async Task Main()
        {
            AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);

            St.Start();
            await DoWork();
            St.Stop();
            Console.WriteLine("Time -" + St.Elapsed);
        }


        static string GetNextUrl()
        {
            if (_urlindex >= Urls.Length)
            {
                return null;
            }

            var url = Urls[_urlindex];
            _urlindex++;

            return url;
        }

        private static async Task DoWorkHttp(string url)
        {
            //using (var handler = new SocketsHttpHandler())
            using (var handler = new HttpClientHandler())
            using (var http = new HttpClient(handler))
            {
                try
                {
                    await http.GetStringAsync("http://" + url);
                }
                catch (Exception)
                {
                    // ignored
                }
            }
        }



        static async Task DoWork()
        {
            var listWorkers = new List<Task>();
            for (int i = 0; i < MaxTasks; i++)
            {
                var url = GetNextUrl();
                if (url == null) { break; }
                var work = DoWorkHttp(url);
                listWorkers.Add(work);
            }

            int countFinished = 0;

            while (listWorkers.Count > 0)
            {
                //Wait one task
                var network = await Task.WhenAny(listWorkers);
                listWorkers.Remove(network);
                await network;

                countFinished++;

                //ShowResult
                Console.WriteLine($"CountFinished -{countFinished}, Time -{St.Elapsed}, q/s -{ countFinished/St.Elapsed.TotalSeconds }");

                //Add new Task
                var url = GetNextUrl();
                if (url != null)
                {
                    var work = DoWorkHttp(url);
                    listWorkers.Add(work);
                }

            }


        }

    }
}



if Enabled  AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);
**q/s ~60**

if Disabled  AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);
**q/s ~6**

**Comment by logwlogw** [comment]
Also my testing 1000urls.txt list 

for Example
google.com
youtube.com
facebook.com
baidu.com
wikipedia.org
qq.com
taobao.com

without www, for many Http Redirect

**Comment by rmkerr** [comment]
That's a really interesting result, and is worth investigating in more depth. Thanks for sharing this.

To summarize: There appears to be a significant perf regression (10x) between SocketsHttpHandler and CurlHandler when following redirects. I'm going to add this to the 3.0 milestone, so we can get it fixed for the next release.

**Comment by rmkerr** [comment]
Does this still repro when you use a static HttpClient rather than instantiating one for every request?

**Comment by logwlogw** [comment]
Yes, 
but for testting  with static client, i add for close sockets
--------------------------------------------------
 GlobalHandel.PooledConnectionIdleTimeout=TimeSpan.Zero;
 GlobalHandel.PooledConnectionLifetime = TimeSpan.Zero;
--------------------------------------------------


`using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net.Http;
using System.Threading.Tasks;

namespace PerfTest
{
    class Program
    {
        static readonly int MaxTasks = 100;

        static readonly Stopwatch St = new Stopwatch();

        private static readonly string[] Urls = File.ReadAllLines(@"10000urls.txt");
        private static int _urlindex;

        private static SocketsHttpHandler GlobalHandel;
        private static HttpClient GlobalHttpClient;


        static async Task Main()
        {
            GlobalHandel = new SocketsHttpHandler();

            //Add for Close Sockets;
            GlobalHandel.PooledConnectionIdleTimeout=TimeSpan.Zero;
            GlobalHandel.PooledConnectionLifetime = TimeSpan.Zero;


            GlobalHttpClient = new HttpClient(GlobalHandel);

            // AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);

            St.Start();
            await DoWork();
            St.Stop();
            Console.WriteLine("Time -" + St.Elapsed);
        }


        static string GetNextUrl()
        {
            if (_urlindex >= Urls.Length)
            {
                return null;
            }

            var url = Urls[_urlindex];
            _urlindex++;

            return url;
        }

        private static async Task DoWorkHttp(string url)
        {
            try
            {
                await GlobalHttpClient.GetStringAsync("http://" + url);
            }
            catch (Exception)
            {
                // ignored
            }


        }



        static async Task DoWork()
        {
            var listWorkers = new List<Task>();
            for (int i = 0; i < MaxTasks; i++)
            {
                var url = GetNextUrl();
                if (url == null) { break; }
                var work = DoWorkHttp(url);
                listWorkers.Add(work);
            }

            int countFinished = 0;

            while (listWorkers.Count > 0)
            {
                //Wait one task
                var network = await Task.WhenAny(listWorkers);
                listWorkers.Remove(network);
                await network;

                countFinished++;

                //ShowResult
                Console.WriteLine($"CountFinished -{countFinished}, Time -{St.Elapsed}, q/s -{ countFinished / St.Elapsed.TotalSeconds }");

                //Add new Task
                var url = GetNextUrl();
                if (url != null)
                {
                    var work = DoWorkHttp(url);
                    listWorkers.Add(work);
                }

            }


        }

    }
}
`

**Comment by logwlogw** [comment]
Also I Testing my software and save result every 2 sec.
It is AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);
![image](https://user-images.githubusercontent.com/45608226/49906926-aefcc100-fe7c-11e8-9f12-6f54e9ba55d4.png)


It Is Disabled  AppContext.SetSwitch("System.Net.Http.UseSocketsHttpHandler", false);
![image](https://user-images.githubusercontent.com/45608226/49906981-e66b6d80-fe7c-11e8-85e5-5770027d98ea.png)

other bug wit libcurl handel memory leaks :(

**Comment by logwlogw** [comment]
I think maybe the problem is in SSL

**Comment by logwlogw** [comment]
Yes,  problem is in SSL, for Unix systems

I write custom simple  HttpMessageHandler

Problem with ssl, - CPU 99% Load
-------------------------------
         (socket, stream) = await ConnectHelper.ConnectAsync(host,443, _cancellationToken).ConfigureAwait(false);

            (stream) = await ConnectHelper.EstablishSslConnectionAsync(options, request, stream, _cancellationToken);
-------------------------------

**Comment by rmkerr** [comment]
Thanks for the additional info!

@wfurt, do you know if our SSL performance is significantly worse on Unix?

**Comment by wfurt** [comment]
I don't have any good comparison @rmkerr. There are - at least - two parts. Cost of new ssl handshake and than performance of actual crypto. From the conversation above it is not clear if we use same ciphers and if we use same connection strategy. 
I know it may not matter from external prospective but if for example do more ssl handshake in one case we would see performance impact.

**Comment by geoffkizer** [comment]
@stephentoub was looking at some issues re SSL handshake on Linux. I believe we have some inefficiencies here.

**Comment by stephentoub** [comment]
What OS is this?

And why are you setting the pooled connection timeouts to 0?

We do have a perf issue with creating SSL connections on Linux:
https://github.com/dotnet/corefx/issues/35086
https://github.com/aspnet/AspNetCore/issues/7081

**Comment by davidsh** [comment]
We think these issues are fixed in latest .NET Core 3.0.  Can you please try that?

Latest preview:
https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0-preview-8/
